### ISSUE TITLE
An unnecessary transitive dependency on github:m-demare/hlargs.nvim

### ISSUE THREAD

**Issue by nikololiahim** [BirdeeHub/nixCats-nvim#31]
Hi! I have recently wanted to change my neovim configuration to be more lua-centric (coming from Nixvim), so I have decided to try this project. It's been quite a ride so far! First, I instantiated the baseline flake using the basic template:

```
nix flake init -t github:BirdeeHub/nixCats-nvim
```

Then, after building and running what is supposed to be the barebones Neovim, I have discovered that the nixCats flake depends on `github:m-demare/hlargs.nvim`. The output of of `nix flake metadata`:

```
Resolved URL:  REDACTED
Description:   A Lua-natic's neovim flake, with extra cats! nixCats!
Path:          /nix/store/7vq6v0h97kl8mw4x9drjcwmzx256kb3x-source
Revision:      edd64d723b61ca9a42777c9d4514f0d1a5578efc-dirty
Last modified: 2024-07-03 00:21:15
Inputs:
├───nixCats: github:BirdeeHub/nixCats-nvim/5eb4a84b31be655fe5a60248e061399f276aaabb?narHash=sha256-ml8HQi/UYHuhpdZQZuNfriFXmziZoQIB/QvlgfMQ1oo%3D (2024-06-28 17:05:47)
│   ├───nixpkgs follows input 'nixpkgs'
│   └───plugins-hlargs: github:m-demare/hlargs.nvim/30fe1b3de2b7614f061be4fc9c71984a2b87e50a?narHash=sha256-V3XG1SCBz%2BEvObW7JSNIUxGZLv6zoRXyap7nRETmUA8%3D (2024-05-18 17:47:37) # weird!
└───nixpkgs: github:nixos/nixpkgs/0f56e3221392452f2c38e3ddf2eba03abda6bf47?narHash=sha256-n0r/%2BL73jFW4Y%2BsCUerulR7VcVADnfcXdzRs7PaiIqw%3D (2024-07-02 06:03:53)

```
The hlargs plugin bothered me at first, because I didn't install it and it is somehow in my flake's inputs. Then I looked at the flake in this repo and realised that this plugin is a transitive dependency of my flake.
The issue is that I can't even use this plugin, but when I run `nix flake update` it gets updated in the flake.lock file. It bothers me, because I don't want to depend on plugins I don't use. 

On the other hand, I get it that it is there in the main flake to demonstrate how to install plugins that are not in the main nixpkgs. So, is it possible to somehow exclude this plugin from the flake that I base my own flake on?

**Comment by BirdeeHub** [comment]
Its only actually included in your config if you add it to the list of plugins you install but it is fetched still even if it isnt added. It is quite small though and it only fetches 1 commit not the whole repository and all its commits. If you dont include it, it never gets ran, but yeah it does get fetched.

You can actually use it. You can merge the exported dependency overlays from other nixCats flakes with your own like this: [here is how you would use the plugins imported in this manner by another nixCats config](https://github.com/BirdeeHub/birdeeSystems/blob/9b611159e957c18f71545e406952e6ac7ea7a5da/common/birdeevim/default.nix#L9-L20) This is my config, which does actually use the hlargs source from the nixCats repo.

Obviously, if you dont use the plugin though, it does feel bad to import something you dont use.

Unfortunately you cant remove inputs from flakes that you import. You can override them to different versions using the follows mechanism, but not remove.

Eelco is supposedly working on lazy trees for flakes, which would stop it from fetching if its not used.

I am open to removing it. If I was to remove it, I would simply comment it out so it could still be there for tutorial purposes, just like how importing the neovim nightly flake is demonstrated.

But yeah, it has been a small thing, but I suppose removing it does make sense. At the same time I don't like losing the tutorial value it provides.

Basically, it hasnt been a problem so far but sure, it makes some sense to remove it, because importers of nixCats cannot get rid of it themselves.

**Comment by BirdeeHub** [comment]
Basically, I dont like losing the tutorial value, but I think it is probably worth it to remove it despite losing the tutorial value

**Comment by BirdeeHub** [comment]
done :)

**Comment by BirdeeHub** [comment]
Ok well, turns out someone other than me was actually using it but, like, they can copy paste the input thats fine XD

**Comment by nikololiahim** [comment]
Great, thank you! I will update the flake when I get the chance. I have seen the removed input, so I'll close this. I may also submit a few PRs to the README and docs, because they are... all over the place. 

Once again, thank you for actively maintaining this project! I was able to successfully install plugins with Nix and configure them with Lua, just like anywhere else.

**Comment by BirdeeHub** [comment]
I would love that. Docs are hard for me, and the readme is pretty bad XD There is a lot of stuff people dont know about nix and neovim, and a lot of levers that I gave people to pass info and dependencies. It was hard enough to just get all the info on the page XD

**Comment by BirdeeHub** [comment]
Everything in the templates is from the nixCats.utils set? So maybe that should be more front and center? But the main thing people interact with is the categoryDefinitions and packageDefinitions sets, so those also needed to be important? There was a lot to document so figuring out how to layout the docs was hard.

**Comment by BirdeeHub** [comment]
also, to be fair, the only thing that isnt docs is the nix/builder and nix/utils directories XD the rest is all docs or example or template. So yeah, the docs are ALL OVER the place XD as in, almost the whole thing XD

**Comment by nikololiahim** [comment]
I was thinking something along the lines of a tutorial where we start with the most barebones, minimalistic flake possible and add stuff to it, explaining each step of the way. Like, for example, how to configure `wrapRc` and `unwrappedCfgPath` to enable the quick feedback loop, say how unwrappedCfgPath is not reproducible on another machine because it requires an absolute path, etc etc.

**Comment by BirdeeHub** [comment]
Thats an interesting idea, but the most barebones flake is like, still a lot of lines due to the outputs section. maybe do that part on its own and then for the rest do it like that and leave that part out?

https://github.com/BirdeeHub/nixCats-nvim/blob/52c9b7e111fc867a575ed80cd205ce466ba2b8a6/flake.nix#L377-L449

Specifically this section is a bit long and hard to minimize and would add a lot of noise to such a tutorial. The rest would work fairly well for a tutorial like this.

my approach was to kinda cover it in sections in nixCatsFlake.txt where the first section is for the inputs, second for overlays, 3rd for categoryDefinitions function, 4th for packageDefinitions function, then 5th for the outputs section linked above?

Then I have a separate nixCatsModules.txt with literally just the options hoping they can match them up with how they are in the flake, which could really use a touch up XD

Then the nixCats global function is in its own file.

Then theres a file for the luaUtils template and one for how to do LSPs in a nice way, and one about a good way to manage custom overlays.

So, if done in 2 or 3 parts doing something like you suggest could work quite well, but there is really a lot to cover.

**Comment by BirdeeHub** [comment]
Im not super picky though as long as it covers all of the info and is readable and not as long as a book.
